diff --git a/dist/parse.js b/dist/parse.js
index 6e74a9d674fdc4d6cebfb571cb92c69bd2918feb..e6193752614031aebf3ea27e062a59b7d1349102 100644
--- a/dist/parse.js
+++ b/dist/parse.js
@@ -105,10 +105,12 @@ export function createMarkdownParser(options = {}) {
                 state.reusableNodes = reusedNodes;
             }
             if (autoClose) {
-                state.markdown = withSpan(tracer, 'comark:autoclose', () => autoCloseMarkdown(state.markdown, {
-                    frontmatter: hasPlugin('frontmatter') && opts.streaming,
-                    syntax: hasPlugin('components'),
-                }));
+                state.markdown = withSpan(tracer, 'comark:autoclose', () => typeof autoClose === 'function'
+                    ? autoClose(state.markdown)
+                    : autoCloseMarkdown(state.markdown, {
+                        frontmatter: hasPlugin('frontmatter') && opts.streaming,
+                        syntax: hasPlugin('components'),
+                    }));
             }
             for (const plugin of plugins) {
                 if (!plugin.pre)
diff --git a/dist/types.d.ts b/dist/types.d.ts
index 9f0c4f8dbb9be1da3e72fb33be4e3db6c6c1975b..c5404b31934a3a9c311657a4e8b22a645c24bd01 100644
--- a/dist/types.d.ts
+++ b/dist/types.d.ts
@@ -334,6 +334,7 @@ export type MergePluginFrontmatter<TPlugins extends readonly unknown[]> = TPlugi
 export type ResolvedMeta<T> = [keyof T] extends [never] ? Record<string, any> : T & Record<string, unknown>;
 export type ResolvedFrontmatter<T> = [keyof T] extends [never] ? Record<string, any> : T & Record<string, unknown>;
 export type ComponentManifest = (name: string) => unknown | Promise<unknown> | undefined | null;
+export type AutoCloseFunction = (markdown: string) => string;
 export interface ComarkContextProvider {
     components: Record<string, any>;
     componentManifest: ComponentManifest;
@@ -381,7 +382,7 @@ export interface ParserOptions<TPlugins extends readonly ComarkPlugin<any, any>[
      * Whether to automatically close unclosed markdown and Comark components.
      * @default true
      */
-    autoClose?: boolean;
+    autoClose?: boolean | AutoCloseFunction;
     /**
      * @deprecated Use `registerDefaultPlugins: false` and register plugins explicitly
      * (include `html()` from `comark/plugins/html` only when you need HTML parsing).
